File: tests/specs_checkout.py
"""Cross-repo guard: skip locally, fail loudly in CI.

Several Chamber conformance tests read fixtures from Noema-Specs checked out
beside this repo and guard themselves with `pytest.skip` so a bare clone still
runs. That is right locally and wrong in CI, where the checkout is supposed to
be there: a missing one turns the suite green while the ADR-005 conformance
checks quietly stop running.

Measured before this existed: with Noema-Specs absent the Python suite reports
`531 passed, 7 skipped` and exits zero. Three of those skips are the cross-repo
conformance tests -- seed-replay equivalence against the published fixtures and
the event-catalog check. The other four are legitimately optional (Postgres, and
the cohort E2E that needs an official-client checkout).

The Worker suite had the same hole and closed it in Noema #609; this is the
Python half of that fix. CI clones Noema-Specs pinned to `spec-compat.json`
`specs.commit`, so a lagging pin can also make an artifact vanish -- which is
exactly how the RFC-0129 conformance suite would have skipped in CI (#608).
"""

from __future__ import annotations

import os
from pathlib import Path

import pytest

__all__ = ["require_specs"]


def require_specs(path: Path, what: str) -> None:
    """Skip when `path` is missing locally; fail when missing under CI.

    `what` names the artifact in the failure so the cause is actionable
    rather than a bare path.
    """
    if path.exists():
        return
    message = (
        f"Noema-Specs artifact missing: {what} at {path}. "
        "CI clones Noema-Specs at spec-compat.json `specs.commit`, not main. "
        "If the artifact landed in Specs after that commit, advance the pin; "
        "otherwise fix the path. Skipping here would pass while checking nothing."
    )
    if os.environ.get("CI"):
        pytest.fail(message, pytrace=False)
    pytest.skip(message)

File: workers/noema/test/specs-checkout.ts
import { existsSync } from "node:fs";

/**
 * Cross-repo guard: skip locally, fail loudly in CI.
 *
 * Several suites read fixtures from Noema-Specs checked out beside this repo
 * and guard themselves with `skipIf(!have)` so a bare clone still runs. That is
 * right locally and wrong in CI, where the artifact is supposed to be there:
 * a missing one turns the suite green while checking nothing.
 *
 * This is not hypothetical. Noema CI clones Noema-Specs pinned to
 * `spec-compat.json` `specs.commit`, not main. When RFC-0129 landed in Specs,
 * that pin still named a commit from before it, so the RFC-0129 conformance
 * fixtures did not exist in the CI checkout and its whole suite would have
 * skipped — passing while proving nothing (Noema #608). The same trap is open
 * to every cross-repo suite whenever the pin lags an artifact it needs.
 *
 * In CI a missing artifact now names itself and the likely cause. Locally the
 * skip is preserved, so a bare clone still runs the suite.
 */
export function haveSpecsArtifacts(...paths: string[]): boolean {
  const missing = paths.filter((p) => !existsSync(p));
  if (missing.length === 0) return true;
  if (process.env.CI) {
    throw new Error(
      [
        `Noema-Specs artifact missing in CI: ${missing.join(", ")}.`,
        "CI clones Noema-Specs at spec-compat.json `specs.commit`, not main.",
        "If the artifact landed in Specs after that commit, advance the pin;",
        "otherwise fix the path. Skipping here would pass while checking nothing.",
      ].join(" "),
    );
  }
  return false;
}
